### PR TITLE
issue: 1284069 Modify ib device mapping

### PR DESCRIPTION
**Pull request by igor-ivanov**

### CODE REVIEW

**Comment by igor-ivanov** [comment]
bot:retest

**Comment by igor-ivanov** [comment]
bot:retest

File: src/vma/dev/net_device_val.cpp [hunk 1]
```diff
@@ -136,11 +130,132 @@ void ring_alloc_logic_attr::set_user_id_key(uint64_t user_id_key)
 	}
 }
 
-net_device_val::net_device_val(transport_type_t transport_type) : m_if_idx(0), m_local_addr(0),
-m_netmask(0), m_mtu(0), m_state(INVALID), m_p_L2_addr(NULL), m_p_br_addr(NULL),
-m_transport_type(transport_type),  m_lock("net_device_val lock"), m_bond(NO_BOND),
-m_bond_xmit_hash_policy(XHP_LAYER_2), m_bond_fail_over_mac(0)
+net_device_val::net_device_val(void *desc) : m_lock("net_device_val lock")
 {
```
Comment (liranoz12) [inline src/vma/dev/net_device_val.cpp]: Why did you change the init method ?

File: src/vma/dev/net_device_val.cpp [hunk 2]
```diff
@@ -136,11 +130,132 @@ void ring_alloc_logic_attr::set_user_id_key(uint64_t user_id_key)
...
+		return ;
+	}
+
+	nd_logdbg("Checking if can offload on interface '%s' (addr=%d.%d.%d.%d, flags=%X)",
+			ifa->ifa_name, NIPQUAD(((struct sockaddr_in *)ifa->ifa_addr)->sin_addr.s_addr), ifa->ifa_flags);
+
+	cma_id = NULL;
+	IF_RDMACM_FAILURE(rdma_create_id(p_cma_event_channel, &cma_id, NULL, RDMA_PS_UDP)) { // UDP vs IP_OVER_IB?
+		nd_logerr("Failed in rdma_create_id (RDMA_PS_UDP) (errno=%d %m)", errno);
+		return ;
```
Comment (liranoz12) [inline src/vma/dev/net_device_val.cpp]: We need to destroy the event channel here

File: src/vma/dev/net_device_val.cpp [hunk 3]
```diff
@@ -811,10 +889,8 @@ void net_device_val::unregister_to_ibverbs_events(event_handler_ibverbs *handler
 	}
 }
 
-void net_device_val_eth::configure(struct ifaddrs* ifa, struct rdma_cm_id* cma_id)
+void net_device_val_eth::configure(struct ifaddrs* ifa)
 {
```
Comment (liranoz12) [inline src/vma/dev/net_device_val.cpp]: Do we need this function now ? can it be merged to the constructor ?

File: src/vma/dev/net_device_val.cpp [hunk 4]
```diff
@@ -222,9 +222,7 @@ net_device_val::net_device_val(void *desc) : m_lock("net_device_val lock")
 	m_local_addr    = ((struct sockaddr_in *)ifa->ifa_addr)->sin_addr.s_addr;
 	m_netmask       = ((struct sockaddr_in *)ifa->ifa_netmask)->sin_addr.s_addr;
 
-	char base_ifname[IFNAMSIZ];
-	get_base_interface_name((const char*)(ifa->ifa_name), base_ifname, sizeof(base_ifname));
-	if (check_device_exist(base_ifname, BOND_DEVICE_FILE)) {
+	if (check_device_exist(m_base_name, BOND_DEVICE_FILE)) {
 		// this is a bond interface (or a vlan/alias over bond), find the slaves
```
Comment (liranoz12) [inline src/vma/dev/net_device_val.cpp]: There are 3 more locations which we can replace with the new call : 
 
> [lirano@r-aa-apollo03 10:32:56 libvma]$ grep -nir "char base_ifname" src/vma/dev/net_device_val.cpp -A 1
380:            char base_ifname[IFNAMSIZ];
381-            if (get_base_interface_name((const char*)m_slaves[i]->if_name, base_ifname, sizeof(base_ifname))) {
--
1118:   char base_ifname[IFNAMSIZ];
1119-   get_base_interface_name((const char*)(ifa->ifa_name), base_ifname, sizeof(base_ifname));
--
1242:   char base_ifname[IFNAMSIZ];
1243-   get_base_interface_name((const char*)(ifname), base_ifname, sizeof(base_ifname));

File: src/vma/proto/neighbour.cpp [hunk 5]
```diff
@@ -2007,12 +2012,7 @@ int neigh_ib::find_pd()
 {
 	neigh_logdbg("");
 
-	if (m_cma_id->verbs == NULL) {
-		neigh_logdbg("m_cma_id->verbs is NULL");
-		return -1;
-	}
-	ib_ctx_handler* ib_ctx_h = g_p_ib_ctx_handler_collection->get_ib_ctx(
-			m_cma_id->verbs);
+	ib_ctx_handler* ib_ctx_h = g_p_ib_ctx_handler_collection->get_ib_ctx(m_p_dev->get_ifname_link());
```
Comment (liranoz12) [inline src/vma/proto/neighbour.cpp]: What about ib_ctx_handler_collection::get_ib_ctx(struct ibv_context* p_ibv_context) ? can we remove it too ?

File: src/vma/dev/ib_ctx_handler.cpp [hunk 6]
```diff
@@ -55,7 +55,7 @@
 #define ibch_logfuncall         __log_info_funcall
 
 
-ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t ctx_time_converter_mode) :
+ib_ctx_handler::ib_ctx_handler(void *desc, ts_conversion_mode_t ctx_time_converter_mode) :
```
Comment (liranoz12) [inline src/vma/dev/ib_ctx_handler.cpp]: Why void type ?

File: src/vma/dev/ib_ctx_handler_collection.h [hunk 7]
```diff
@@ -46,16 +46,18 @@ class ib_ctx_handler_collection
...
-			uint64_t access);
+	size_t get_num_devices() {return m_n_num_devices; };
+	size_t mem_reg_on_all_devices(void* addr, size_t length,
+			ibv_mr** mr_array, size_t mr_array_sz, uint64_t access);
 	ts_conversion_mode_t    get_ctx_time_conversion_mode();
 
 private:
+	void update_tbl();
+	void print_val_tbl();
+
```
Comment (liranoz12) [inline src/vma/dev/ib_ctx_handler_collection.h]: I dont like this renaming - The api is for classes which inherint from netlink_socket_mgr but this is not the case here.

File: src/vma/dev/net_device_val.cpp [hunk 8]
```diff
@@ -242,7 +245,7 @@ net_device_val::net_device_val(void *desc) : m_lock("net_device_val lock")
 			nd_logdbg("Blocking offload: No verbs context in cma_id on interfaces ('%s')", get_ifname());
 			goto err;
 		}
-		ib_ctx = g_p_ib_ctx_handler_collection->get_ib_ctx(cma_id->verbs);
+		ib_ctx = g_p_ib_ctx_handler_collection->get_ib_ctx(get_ifname_link());
```
Comment (liranoz12) [inline src/vma/dev/net_device_val.cpp]: This can cause issues with bond devices

File: src/vma/proto/neighbour.cpp (hunk 5 above)
Comment (liranoz12) [inline src/vma/proto/neighbour.cpp]: This can cause issue with bond devices

File: src/vma/dev/ib_ctx_handler.cpp (hunk 6 above)
Comment (liranoz12) [inline src/vma/dev/ib_ctx_handler.cpp]: Please replace void param as for netdev

File: src/vma/dev/time_converter.cpp [hunk 9]
```diff
@@ -87,19 +87,25 @@ uint32_t time_converter::get_single_converter_status(struct ibv_context* ctx) {
...
-                        "status for all devices [%d], ibv_context_list = %p\n", num_devices, ibv_context_list);
+                        "status for all devices [%d], ibv_dev_list = %p\n", num_devices, ibv_dev_list);
 
 	if (safe_mce_sys().hw_ts_conversion_mode != TS_CONVERSION_MODE_DISABLE){
 		devs_status = IBV_EXP_QUERY_DEVICE_SUPPORTED | IBV_EXP_QUERY_VALUES_SUPPORTED;
 		for (int i = 0; i < num_devices; i++) {
-			devs_status &= get_single_converter_status(ibv_context_list[i]);
+			struct ibv_context *ibv_ctx = ibv_open_device(ibv_dev_list[i]);
+			if (ibv_ctx == NULL) {
+				ibchtc_logdbg("ibv_ctx is invalid");
```
Comment (liranoz12) [inline src/vma/dev/time_converter.cpp]: Do we expect this failure to happend ? why not error ?

File: src/vma/dev/ib_ctx_handler_collection.cpp [hunk 10]
```diff
@@ -144,14 +141,30 @@ ib_ctx_handler* ib_ctx_handler_collection::get_ib_ctx(struct ibv_context* p_ibv_
...
+			return NULL;
+		}
+		ifa_name = (const char *)slave.ifa_name;
+	} else if (check_device_exist(ifa_name, BOND_DEVICE_FILE)) {
+		if (!get_bond_active_slave_name(ifa_name, active_slave, sizeof(active_slave))) {
+			return NULL;
+		}
+		ifa_name = (const char *)active_slave;
+	}
+
```
Comment (liranoz12) [inline src/vma/dev/ib_ctx_handler_collection.cpp]: This will fix the issue with the bond...

File: src/vma/dev/allocator.cpp [hunk 11]
```diff
@@ -51,6 +52,34 @@ vma_allocator::vma_allocator() :
 	m_contig_access_mr = 0;
 #endif
 
+	__log_info_dbg("Done");
+}
+
```
Comment (liranoz12) [inline src/vma/dev/allocator.cpp]: @rafiw Please have a look here

File: src/vma/dev/ib_ctx_handler.cpp [hunk 12]
```diff
@@ -166,49 +193,114 @@ ib_ctx_handler::~ib_ctx_handler() {
...
+		}
 	}
-	IF_VERBS_FAILURE(ibv_dereg_mr(mr)) {
-		ibch_logerr("failed de-registering a memory region "
-				"(errno=%d %m)", errno);
-	} ENDIF_VERBS_FAILURE;
-	VALGRIND_MAKE_MEM_UNDEFINED(mr, sizeof(ibv_mr));
+}
+
+struct ibv_mr* ib_ctx_handler::get_mem_reg(uint32_t lkey)
```
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.cpp]: consider the name get_mr?

File: src/vma/dev/ib_ctx_handler.cpp [hunk 13]
```diff
@@ -136,24 +143,44 @@ ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t ctx
 #endif // DEFINED_IBV_EXP_CQ_TIMESTAMP
 	// update device memory capabilities
 	m_on_device_memory = vma_ibv_dm_size(m_p_ibv_device_attr);
-	ibch_logdbg("ibv device '%s' [%p] has %d port%s. Vendor Part Id: %d, "
```
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.cpp]: don;t you want to leave that print?

File: src/vma/dev/ib_ctx_handler.cpp [hunk 14]
```diff
@@ -136,24 +143,44 @@ ib_ctx_handler::ib_ctx_handler(struct ibv_context* ctx, ts_conversion_mode_t ctx
...
-		    m_p_ibv_device_attr->fw_ver, m_p_ibv_device_attr->max_qp_wr,
-		    m_p_ibv_device, m_on_device_memory);
 
 	g_p_event_handler_manager->register_ibverbs_event(m_p_ibv_context->async_fd,
 						this, m_p_ibv_context, 0);
+
+	return;
+
+err:
+	if (m_p_ibv_device_attr) {
```
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.cpp]: no need to free resources destructor will delete all of them

File: src/vma/dev/ib_ctx_handler.cpp [hunk 15]
```diff
@@ -166,49 +193,114 @@ ib_ctx_handler::~ib_ctx_handler() {
...
 }
 
-void ib_ctx_handler::mem_dereg(ibv_mr *mr)
+void ib_ctx_handler::mem_dereg(uint32_t lkey)
 {
-	if (is_removed()) {
-		return;
+	if (!is_removed()) {
+		mr_map_lkey_t::iterator iter = m_mr_map_lkey.find(lkey);
+		if (iter != m_mr_map_lkey.end()) {
```
Comment (rafiw) [inline src/vma/dev/ib_ctx_handler.cpp]: consider adding debug\warn if not found?

File: src/vma/dev/ib_ctx_handler.cpp (hunk 6 above)
Comment (igor-ivanov) [reply src/vma/dev/ib_ctx_handler.cpp]: ok

File: src/vma/dev/allocator.cpp [hunk 16]
```diff
@@ -36,11 +36,12 @@
...
 #define MODULE_NAME	"allocator"
 
-vma_allocator::vma_allocator() :
-		m_mr_list(NULL),
-		m_mr_list_len(0),
-		m_shmid(-1),
-		m_data_block(NULL) {
+vma_allocator::vma_allocator()
+{
+	__log_info_dbg("");
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/allocator.cpp]: agree, initialize list requires keeping right order as in class declaration.

File: src/vma/dev/allocator.h [hunk 17]
```diff
@@ -38,7 +38,7 @@
 
 
 class ib_ctx_handler;
-
+typedef std::tr1::unordered_map<ib_ctx_handler*, uint32_t> lkey_map_ib_ctx_map_t;
```
[the remark this answers is not in the document]
Comment (igor-ivanov) [reply src/vma/dev/allocator.h]: it could be considered for future performance investigation

File: src/vma/dev/ib_ctx_handler.cpp (hunk 15 above)
Comment (igor-ivanov) [reply src/vma/dev/ib_ctx_handler.cpp]: it might be usual case during plugout

**Comment by liranoz12** [comment]
Duplicated by #619
